### PR TITLE
Fix failing CI test - loosen requirements in a too demanding test

### PR DESCRIPTION
**Pull request by tomob**
Signed-off-by: Tomasz Baranski <tbaransk@redhat.com>

**What this PR does / why we need it**:

The failing test suggest a better way to get DV's StorageClass. This PR makes the VolumeSnapshotStatus to be calculated without waiting for the underlying PVC.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by rmohr** [comment]
@tomob looks like some unit tests fail now. Also the other lanes don't seem yet to be happy about the change.

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 1]
```diff
@@ -686,24 +688,66 @@ func (ctrl *VMSnapshotController) getVolumeStorageClassForVolume(namespace strin
...
+	if err != nil {
+		return "", err
+	}
+
+	var dv *cdiv1.DataVolume
+	if exists {
+		dv = obj.(*cdiv1.DataVolume)
+		if dv.Spec.PVC == nil || dv.Spec.PVC.StorageClassName == nil {
+			log.Log.V(3).Infof("DV [%s] not found", key)
+			return "", fmt.Errorf("DV '%s' not found", key)
```
Comment (rmohr) [inline pkg/virt-controller/watch/snapshot/snapshot.go:716]: Could you explain what this `not found` means? It looks to me like you found it in this code branch.

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 2]
```diff
@@ -686,24 +688,66 @@ func (ctrl *VMSnapshotController) getVolumeStorageClassForVolume(namespace strin
...
+			if err != nil || !exists {
+				continue
+			}
+
+			vm, ok := storeObj.(*kubevirtv1.VirtualMachine)
+			if !ok {
+				continue
+			}
+
+			for _, dvTemplate := range vm.Spec.DataVolumeTemplates {
```
Comment (rmohr) [inline pkg/virt-controller/watch/snapshot/snapshot.go:742]: Could you explain why one has to look up a VM? If there is no real DV, why do I have to look up a storage class for it and not just do nothing for the DV?

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 1 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/snapshot.go:716]: nil `dv.Spec.PVC.StorageClassName` is okay here.  Either means use the default storage class or no storage class (statically allocated PVs)

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 2 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/snapshot.go:742]: Could wait for the DV to be created but why wait if the info is in the template?

Comment (tomob) [reply pkg/virt-controller/watch/snapshot/snapshot.go:742]: This point is reached not only when the DV is not found, but also when the StorageClass for the DV has not been found. Previously we waited for the PVC to be ready to read its StorageClass, but this has a side effect: the status is (possibly) wrongly set to `false` and later updated to a whatever it should be (true or false).

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 1 above)
Comment (rmohr) [reply pkg/virt-controller/watch/snapshot/snapshot.go:716]: This returned error still looks suspicious. As far as I can read the code, the PVC may not be found, but it is still misleading, because why would a DV or PVC not be found if there is no StorageClassName set?

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 2 above)
Comment (rmohr) [reply pkg/virt-controller/watch/snapshot/snapshot.go:742]: Let me rephrase my question: If the DV and PVC does not exist, can we ever reach this point?

Comment (tomob) [reply pkg/virt-controller/watch/snapshot/snapshot.go:742]: No, we can't.

Let me refactor the code to make it clearer. Also, I seem to have lost the "last resort" case of looking at the PVC's StorageClass.

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 1 above)
Comment (tomob) [reply pkg/virt-controller/watch/snapshot/snapshot.go:716]: You're right, this is misleading. Refactoring...

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 3]
```diff
@@ -686,24 +688,77 @@ func (ctrl *VMSnapshotController) getVolumeStorageClassForVolume(namespace strin
...
+				if dvTemplate.Name == dvName && dvTemplate.Spec.PVC != nil {
+					return *dvTemplate.Spec.PVC.StorageClassName, nil
```
Comment (mhenriks) [inline pkg/virt-controller/watch/snapshot/snapshot.go:738]: It is legit for `dvTemplate.Spec.PVC.StorageClassName` to be nil

Comment (tomob) [reply pkg/virt-controller/watch/snapshot/snapshot.go:738]: OK

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 4]
```diff
@@ -686,24 +688,77 @@ func (ctrl *VMSnapshotController) getVolumeStorageClassForVolume(namespace strin
...
 	}
 
-	return "", fmt.Errorf("Volume type does not suport snapshots")
+	if !exists {
+		log.Log.V(3).Infof("PVC not in cache [%s]", pvcKey)
+		return "", fmt.Errorf("PVC for the DataVolume `%s` not found", dvName)
+	}
+
+	pvc := obj.(*corev1.PersistentVolumeClaim)
+	return *pvc.Spec.StorageClassName, nil
```
Comment (mhenriks) [inline pkg/virt-controller/watch/snapshot/snapshot.go:761]: `pvc.Spec.StorageClassName` may be nil and that is okay

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 5]
```diff
@@ -686,24 +688,82 @@ func (ctrl *VMSnapshotController) getVolumeStorageClassForVolume(namespace strin
...
+		return "", fmt.Errorf("PVC for the DataVolume `%s` not found", dvName)
+	}
+
+	pvc := obj.(*corev1.PersistentVolumeClaim)
+	if pvc.Spec.StorageClassName != nil {
+		return *pvc.Spec.StorageClassName, nil
+	}
+
+	log.Log.V(3).Info("PVC has no StorageClassName")
+	return "", fmt.Errorf("PVC has no StorageClassName")
```
Comment (mhenriks) [inline pkg/virt-controller/watch/snapshot/snapshot.go:766]: This is not an error

Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/snapshot.go:766]: Treat it as a SC that does not support snapshots

Comment (tomob) [reply pkg/virt-controller/watch/snapshot/snapshot.go]: Done

**Comment by rmohr** [comment]
This one looks interesting, is this new or was it there before: https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4462/pull-kubevirt-e2e-k8s-1.17/1328617864038453248 ?

**Comment by tomob** [comment]
> This one looks interesting, is this new or was it there before: https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4462/pull-kubevirt-e2e-k8s-1.17/1328617864038453248 ?

I tried it a couple of times. I cannot reproduce the test.

```
> export | grep KUBE 
KUBECONFIG /home/tbaransk/work/repos/cnv/kubevirt/_ci-configs/k8s-1.17/…
KUBEVIRT_PROVIDER k8s-1.17

> make functest 
[...]
Running Suite: Tests Suite
==========================
Random Seed: 1605620550
Will run 1 of 694 specs

JUnit report was created: /home/tbaransk/work/repos/cnv/kubevirt/_out/artifacts/junit.functest.xml

Ran 1 of 694 Specs in 23.604 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 693 Skipped
PASS | FOCUSED

Ginkgo ran 1 suite in 25.448289919s
Test Suite Passed
Detected Programmatic Focus - setting exit status to 197
```
